walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,119 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendeesInRequest = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    ArrayList<TimeRange> eventsTimeRangeList = new ArrayList<>();
+    ArrayList<TimeRange> optionalsEventTimeRangeList = new ArrayList<>();
+    
+    // Returns no options if duration is longer than a day.
+    if(request.getDuration() >  TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList();
+    }
+
+    if(events.size() > 0) {
+
+        // Adds TimeRanges for events that required attendees will be attending.
+        for (Event event : events) {
+            for(String attendee : event.getAttendees()) {
+                if(attendeesInRequest.contains(attendee)) {
+                    eventsTimeRangeList.add(event.getWhen());
+                }
+            }
+        }
+
+        // Adds TimeRanges for events that optional attendees will be attending.
+        if(optionalAttendees.size() > 0) {
+            for(Event event : events) {
+                for(String optionalAttendee : event.getAttendees()) {
+                    if(optionalAttendees.contains(optionalAttendee)) {
+                        optionalsEventTimeRangeList.add(event.getWhen());
+                    }
+                }
+            }
+        }
+
+        // If no required or optional attendees have any other events, return whole day.
+        if(eventsTimeRangeList.size() == 0 && optionalsEventTimeRangeList.size() == 0) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+
+        // Store available TimeRange slots for optional attendees.
+        ArrayList<TimeRange> optionalsSlots = new ArrayList<>();
+        if(optionalsEventTimeRangeList.size() > 0) {       
+            optionalsSlots = findSlots(optionalsEventTimeRangeList, request);
+        }
+
+        // Store available TimeRange slots for required attendees.
+        ArrayList<TimeRange> mandatorySlots = new ArrayList<>();
+        if(eventsTimeRangeList.size() > 0) {
+            mandatorySlots = findSlots(eventsTimeRangeList, request);
+        }
+
+        // Finds slots that are common to both requuired and optional attendees and returns them
+        // if there are any.
+        ArrayList<TimeRange> combinedSlots = new ArrayList<>();
+        if(optionalsSlots.size() > 0) {
+            for(TimeRange optionalSlot : optionalsSlots) {
```

Comment:
(Optional) This part is O(n^2). You can actually reduce it to O(n) by compare sorted TimeRange().
Think about it but change is optional.

Reply:
i've implemented another approach to eliminate the need for this comparison here so as to avoid the o(n^2)